library/core/src/sync/atomic.rs

```diff
@@ -24,26 +24,37 @@
 //!
 //! ## Memory model for atomic accesses
 //!
-//! Rust atomics currently follow the same rules as [C++20 atomics][cpp], specifically `atomic_ref`.
-//! Basically, creating a *shared reference* to one of the Rust atomic types corresponds to creating
-//! an `atomic_ref` in C++; the `atomic_ref` is destroyed when the lifetime of the shared reference
-//! ends. A Rust atomic type that is exclusively owned or behind a mutable reference does *not*
-//! correspond to an “atomic object” in C++, since the underlying primitive can be mutably accessed,
-//! for example with `get_mut`, to perform non-atomic operations.
+//! Rust atomics currently follow the same rules as [C++20 atomics][cpp], specifically the rules
+//! from the [`intro.races`][cpp-intro.races] section, without the "consume" memory ordering. Since
+//! C++ uses an object-based memory model whereas Rust is access-based, a bit of translation work
+//! has to be done to apply the C++ rules to Rust: whenever C++ talks about "the value of an
+//! object", we understand that to mean the resulting bytes obtained when doing a read. When the C++
+//! standard talks about "the value of an atomic object", this refers to the result of doing an
+//! atomic load (via the operations provided in this module). A "modification of an atomic object"
+//! refers to an atomic store.
+//!
+//! The end result is *almost* equivalent to saying that creating a *shared reference* to one of the
+//! Rust atomic types corresponds to creating an `atomic_ref` in C++, with the `atomic_ref` being
+//! destroyed when the lifetime of the shared reference ends. The main difference is that Rust
+//! permits concurrent atomic and non-atomic reads to the same memory as those cause no issue in the
+//! C++ memory model, they are just forbidden in C++ because memory is partitioned into "atomic
+//! objects" and "non-atomic objects" (with `atomic_ref` temporarily converting a non-atomic object
+//! into an atomic object).
+//!
+//! That said, Rust *does* inherit the C++ limitation that non-synchronized atomic accesses may not
+//! partially overlap: they must be either disjoint or access the exact same memory. This in
+//! particular rules out non-synchronized differently-sized accesses to the same data.
```

Comment:
Maybe I'm misunderstanding but this last sentence sounds off to me. AFAIK, there's nothing prohibiting e.g. concurrent `ptr::read::<u32>` and `ptr::read<u64>` from the same address? I think both of these sentences probably limit themselves to just atomic reads, but that seems like it prevents one of goals from the UCG issue:

> First of all it means replacing a non-atomic access by an atomic one can actually introduce UB (even if there are no alignment issues), which is silly.

from being fully realized, since replacing these reads with an atomic load would then be UB.

At the very least I think it would be good to clarify whether we forbid *non-atomic* differently-sized accesses, and how that interacts with a non-synchronized atomic access (e.g., AtomicU32::load vs. `ptr::read<u64>`).

Reply:
You are right, there are still cases where replacing non-atomic accesses by atomic accesses can introduce UB -- good catch. For mixed-size accesses that will be a lot harder to fix; the academic work I am aware of generally only models atomic accesses that are either disjoint or access identical memory ranges.

Reply:
I think I found a way to deal with mixed-size atomic reads in the Miri data race detector, and that makes me think we can also permit then in the docs.

This should make it actually true that replacing a non-atomic access by an atomic access can never introduce UB. :)

Reply:
Assuming archs (and llvm) play nice.

I know that, on paper, x86 doesn't guarantee atomicy for mixed size accesses without a barrier (in practice, linux would stop working if it actually was a problem), idk about other archs.

Reply:
This is about mixed-sized *reads*.  We already allow them as non-atomics, so they better be well-defined on all arches or we have a larger problem.^^

Non-synchronized mixed-size atomic accesses where at least one access is a write remain forbidden. I think that covers all potential problems.

Reply:
Fair, that should be ok then.